### DISCUSSION TITLE
OAuth2 with react native?

### DISCUSSION THREAD

**Discussion by denisgurev** [pocketbase/pocketbase#2286] [Q&A] [answered]
I'm having difficulties trying to make the social auth to work with react native. After some trial and errors I managed to make it work in the browser following the official example. But I can't figure out how to do the same in my react native app. Has anyone tried to integrate the social auth with react native?

Also @ganigeorgiev  if you need some feedback - Just my 2c but I think it would be better to have the auth settings save action inside each collapsible container. Because PocketBase have so many providers and I'm on a small 15' screen, the "Save" button at the bottom of the page is not immediately visible and it took me a while to understand why my changes were not stored after I collapsed the google panel.

**Comment by ganigeorgiev** [comment, marked as answer]
The OAuth2 mobile flow should be more or less the same as in the browser - https://pocketbase.io/docs/authentication/#web-oauth2-integration. The main difference is that instead of a traditional http redirect, for mobile apps you'll need deeplink (aka. url with eventually some special schema that your mobile app could "catch"). Unfortunately the deeplinks setup will differ depending on what platform you are targeting and the best I could recommend is to search for "deeplink [android/ios] react native" or similar.

Alternatively, if you can wait for v0.15.0 (_most likely it will be released in the next ~1-2 weeks_), you'll be able to authenticate with OAuth2 in a single call without deeplinks or even page reload. This is already implemented in the `develop` branch as part of https://github.com/pocketbase/pocketbase/issues/55#issuecomment-1502980993. For React Native, it could look something like this:

```js
import { Linking } from 'react-native';

const authData = await pb.collection('users').authWithOAuth2({
    provider: 'google',
    urlCallback: async (url) => {
        // open the Google sign-in url in the default browser
        await Linking.openURL(url);
    }
});
```

--- 

About the feedback - earlier today I've actually changed the OAuth2 UI since several other users have reported similar complains. I haven't pushed all the changes yet in the `develop` branch but the OAuth2 settings UI will look like this:

![oauth2_screenshot](https://i.imgur.com/Zat57x3.png)

_(the "Save" action is inside the options panel of each individual OAuth2 provider)_

**Reply by denisgurev** [reply]
Thank you for the detailed and speedy answer!

I wasn't able to make the deeplinks work on android but I'll wait for v0.15.0 to use the simplified version since I'm not in a hurry and in the meantime can work on something else.

Good to know that the ui has changed. I love how it looks in the screenshot ❤️

**Comment by SkillAcademia** [comment]
I am using react native and I am not getting my user back in the authData nor I am getting it in the pocketbase db once the OAuth2 login flow is completed. It just takes me back to my app. I have already set up the redirect uri to my hosted pb as explained in the docs with the above code. Using Pocketbase JS SDK v0.15.2.

**Reply by ganigeorgiev** [reply]
I'm not able to reproduce it locally. Are you sure that the SSE events are properly resolving? Did you check https://github.com/pocketbase/pocketbase/discussions/2484#discussioncomment-6118523?

**Reply by SkillAcademia** [reply]
Yes, I did check everything. Everything works which was previously not working. I am getting `[ClientResponseError 400: Failed to authenticate.]` when I am trying to authenticate now after The Google OAuth flow is completed in the native browser.

**Reply by ganigeorgiev** [reply]
Make sure that your `users` collection doesn't have a required custom field (_or if it has, you need to pass it via the `createData` object prop_) and/or the client request satisfy its Create API rule.

**Reply by SkillAcademia** [reply]
Yep that was the main error. I have some custom api rules which were to be satisfied. `createData` solved the problem. Works flawlessly now. Thank you!

**Reply by SkillAcademia** [reply]
And lastly is there any way to save the name that is returned from the google provider in the database 'name' field. During signup using email and password I am actually also taking a name parameter as input and saving it to database. If there is any snippet that you can provide for the same.

**Reply by ganigeorgiev** [reply]
At the moment you'll have to submit a separate `update()` user request with the custom fields data you want to persist. The OAuth2 data are returned under the `meta` key in the auth response:

```js
const authData = await pb.collection("users").authWithOAuth2({ ... });

if (authData.meta.isNew) {
    await pb.collection("users").update(pb.authStore.model.id, {
        name: authData.meta.name // or any other field from the meta object
    });
}
```
